x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTasksExecutor.java

```diff
@@ -362,4 +397,53 @@ boolean shouldRetry(Exception e) {
 
     }
 
+    static final class IndexMetadataVersionChecker implements BiConsumer<Long, Consumer<Exception>> {
+
+        private static final Logger LOGGER = Loggers.getLogger(IndexMetadataVersionChecker.class);
+
+        private final Client followClient;
+        private final Client leaderClient;
+        private final Index leaderIndex;
+        private final Index followIndex;
+        private final AtomicLong currentIndexMetadataVersion;
+
+        IndexMetadataVersionChecker(Index leaderIndex, Index followIndex, Client followClient, Client leaderClient) {
+            this.followClient = followClient;
+            this.leaderIndex = leaderIndex;
+            this.followIndex = followIndex;
+            this.leaderClient = leaderClient;
+            this.currentIndexMetadataVersion = new AtomicLong();
+        }
+
+        public void accept(Long minimumRequiredIndexMetadataVersion, Consumer<Exception> handler) {
+            if (minimumRequiredIndexMetadataVersion <= currentIndexMetadataVersion.get()) {
+                LOGGER.info("Current index metadata version [{}] is higher or equal than minimum required index metadata version [{}]",
+                    currentIndexMetadataVersion.get(), minimumRequiredIndexMetadataVersion);
+                handler.accept(null);
+            } else {
+                updateMapping(handler);
+            }
+        }
+
+        void updateMapping(Consumer<Exception> handler) {
+            ClusterStateRequest clusterStateRequest = new ClusterStateRequest();
```

Comment:
I think we want a guard otherwise concurrent shard changes responses after a mapping update on the leader are going to all put a new mapping on the follower and maybe we should avoid such an update storm?

Reply:
I've made this change in a separate commit: 9635fe3999322d5083a7b8f281258487a99b3162